// config/gcm.js
var user = require('config/models');
var room = require('config/room');
var request = require('request');

exports.update_token = function (id, gcm_token, callback) {
    // console.log("the id is " + id + " gcm_token is " + gcm_token);
    user.find({ token: id }, function (err, users) {
        if (users.length != 0) {
            var mUser = users[0];
            mUser.gcm_token = gcm_token;
            mUser.save();

            callback({ 'response': "Update Gcm Token Success", 'res': true });

        } else {
            callback({ 'response': "Error while updating gcm token", 'res': false });
        }
    });
}

// exports.call_friends = function (id, roomId, isNavigation, callback) {
//     user.find({ token: id }, function (err, users) {
//         if (users.length != 0) {
//             var mUser = users[0];
            
//             user.find({ username: { $in: mUser.friends } }, function(err, friends) {
//                 if (friends.length != 0) {
//                     callback({ 'response': "Sending notifications to friends", 'res': true });
//                     // console.log(friends);
//                     for (var i = 0; i < friends.length; i++) {
//                         if (friends[i].gcm_token) {
//                             call_friend(mUser.username, friends[i].gcm_token, roomId, isNavigation);
//                         } else {
//                             console.log("Cannot call " + friends[i].username);
//                         }
//                     }
//                 } else {
//                     callback({ 'response': "User has no friends", 'res': false });
//                 }
//             });
//         } else {
//             callback({ 'response': "Cannot find the user", 'res': false });
//         }
//     });
// }

exports.call_friends_by_name = function (mUsername, friends, des, roomId, isNavigation, rate, callback) {
    user.find({username: mUsername}, function(err, users) {
        if (users.length != 0) {
            var mUser = users[0];
            var called = true;
            for (var i = 0; i < friends.length; i++) {
                if (mUser.friends.indexOf(friends[i]) != -1) {
                    console.log("Call " + friends[i]);
                    user.find({username: friends[i]}, function(err, users2) {
                        if (users2.length != 0) {
                            if (users2[0].gcm_token) {
                                console.log(users2[0].gcm_token);
                                call_friend(mUser.username, users2[0].gcm_token, des, roomId, isNavigation, rate);
                                called = true;
                            } else {
                                console.log("Cannot call " + friends[i]);
                            }
                        } else {
                            console.log("Cannot find and ignore " + friends[i]);
                        }
                    });
                } else {
                    console.log(friends[i] + " is not a friend!");
                }
            }
            if (called) callback({ 'response': "Sending notifications to selected friends", 'res': true });
            else callback({ 'response': "Didnot send notifications to any friend", 'res': false });
        } else {
            callback({'response': "Cannot find the user!", 'res': false});
        }
    });
}

var call_friend = function (username, friend_token, des, roomId, isNavigation, rate) {
    console.log(friend_token);
    request({
        url: 'https://gcm-http.googleapis.com/gcm/send',
        headers: {
            'Content-Type': 'application/json',
            Authorization: 'key=AIzaSyCXbMrUxAN_OUiF0kCJeEWPpgoxZw_5Imc'
        },
        method: 'POST',
        json: true,
        body: {
            "data": {
                senderName: username,
                des: des,
                roomId: roomId,
                isNavigation: isNavigation,
                rate: rate,
                time: (new Date()).getTime
            },
            "to": friend_token
        }
    }, function () { });
};
