// config/room.js
var mongoose = require('mongoose'); 
var user = require('config/models'); 

var room_list = {};

// isNavigation is boolean; rate is float
exports.create_room = function(username, des, isNavigation, rate, callback) {
    var roomId = 'a' + Math.floor(Math.random() * 10000000).toString();
    if (room_list[username]) {
        room_list[username] = { "roomId": roomId, 'des': des, 'isNavigation': isNavigation, 'rate': rate, 'free': true };
        callback({ 'response': 'Updated room', 'roomId': roomId });
    } else {
        room_list[username] = { "roomId": roomId, 'des': des, 'isNavigation': isNavigation, 'rate': rate, 'free': true };
        callback({ 'response': 'Created room', 'roomId': roomId });
    }
}

exports.delete_room = function(username, callback) {
    if (room_list[username]) {
        delete room_list[username];
        callback({ 'response': 'Deleted room' });
    } else {
        callback({ 'response': 'Room is not exist!' });
    }
}

exports.enter_room = function(username, callback) {
    if (room_list[username]) {
        room_list[username]['free'] = false;
        callback({ 'response': 'A helper entered room' });
    } else {
        callback({ 'response': 'Room is not exist!' });
    }
}

exports.leave_room = function(username, callback) {
    if (room_list[username]) {
        room_list[username]['free'] = true;
        callback({ 'response': 'The helper left room' });
    } else {
        callback({ 'response': 'Room is not exist!' });
    }
}

exports.get_rooms_by_topic = function(id, topic, callback) {
    var result = [];
    // room_list["John(mock user)"] = { 'des': "SFSU,student", 'isNavigation': true, 'rate': 3.0, 'roomId': '1234', 'free': true };
    // room_list["Mary(mock user)"] = { 'des': "", 'isNavigation': false, 'rate': 4.3, 'roomId': '1235', 'free': true };
    // room_list["name3"] = {'topic': "Cooking", 'rate': "5", 'roomId': '1236', 'free': true};
    
    var count = 0;
    // var isGeneral = (topic === "General");
    var friends = [];
    
    user.find({token: id}, function(err, users) {
        if (users.length != 0) {
            friends = users[0].friends;
            for (var i = 0; i < friends.length; i++) {
                var fname = friends[i];
                if (room_list[fname] && room_list[fname]['free']) {// && (isGeneral || room_list[fname]['topic'] === topic)) {
                    result.push({
                        'username': fname,
                        'des': room_list[fname]['des'],
                        'rate': room_list[fname]['rate'],
                        'isNavigation': room_list[fname]['isNavigation'],
                        'roomId': room_list[fname]['roomId']
                    });
                    count++;
                    console.log("find a friend");
                    if (count == 20) break;
                }
            }
        } else {
            console.log("Cannot find the user!");
        }
        
        for (var username in room_list) {
            if (count == 20) break;
            if (room_list[username]['free']) {// && (isGeneral || room_list[username]['topic'] === topic)) {
                if (friends.indexOf(username) != -1) continue;
                result.push({
                    'username': username,
                    'des': room_list[username]['des'],
                    'rate': room_list[username]['rate'],
                    'isNavigation': room_list[username]['isNavigation'],
                    'roomId': room_list[username]['roomId']
                });
                count++;
            }
        }
        
        callback({ 'response': 'Got List', 'size': count, 'list': JSON.stringify(result) });
    });
    
}
