// config/room.js
var room_list = {};


exports.create_room = function (username, topic, rate, callback) {
    var roomId = Math.floor(Math.random() * 10000000).toString();
    if (room_list[username]) {
        room_list[username] = {"roomId": roomId.toString(), 'topic': topic, 'rate': rate, 'free': true};
        callback({'response': 'Updated room', 'roomId': roomId});
    } else {
        room_list[username] = {"roomId": roomId.toString(), 'topic': topic, 'rate': rate, 'free': true};
        callback({'response': 'Created room', 'roomId': roomId});
    }
}

exports.delete_room = function (username, callback) {
    if (room_list[username]) {
        delete room_list[username];
        callback({'response': 'Deleted room'});
    } else {
        callback({'response': 'Room is not exist!'});
    }
}

exports.enter_room = function (username, callback) {
    if (room_list[username]) {
        room_list[username]['free'] = false;
        callback({'response': 'A helper entered room'});
    } else {
        callback({'response': 'Room is not exist!'});
    }
}

exports.leave_room = function (username, callback) {
    if (room_list[username]) {
        room_list[username]['free'] = true;
        callback({'response': 'The helper left room'});
    } else {
        callback({'response': 'Room is not exist!'});
    }
}

exports.get_rooms_by_topic = function (topic, callback) {
    var result = [];
    // room_list["name1"] = {'topic': "Navigation", 'rate': "3", 'roomId': '1234', 'free': true};
    // room_list["name2"] = {'topic': "Cooking", 'rate': "4", 'roomId': '1235', 'free': true};
    // room_list["name3"] = {'topic': "Cooking", 'rate': "5", 'roomId': '1236', 'free': true};
    var count = 0;
    if (topic === 'General') {
        for (var username in room_list) {
            if (room_list[username]['free']) {
                result.push({
                    'username': username, 
                    'topic': room_list[username]['topic'], 
                    'rate': room_list[username]['rate'], 
                    'roomId': room_list[username]['roomId']
                });
                count++;
                if (count === 20) break;
            }
        }
    } else {
        for (var username in room_list) {
            if (room_list[username]['free'] && room_list[username]['topic'] === topic) {
                result.push({
                    'username': username, 
                    'topic': room_list[username]['topic'], 
                    'rate': room_list[username]['rate'], 
                    'roomId': room_list[username]['roomId']
                });
                count++;
                if (count === 20) break;
            }
        }
    }
    callback({'response':'Got List', 'size':count, 'list':JSON.stringify(result)});
}
