// config/room.js
var mongoose = require('mongoose'); 
var user = require('config/models'); 

var room_list = {};


exports.create_room = function(username, topic, tags, rate, callback) {
    var roomId = Math.floor(Math.random() * 10000000).toString();
    if (room_list[username]) {
        room_list[username] = { "roomId": roomId.toString(), 'topic': topic, 'tags': tags, 'rate': rate, 'free': true };
        callback({ 'response': 'Updated room', 'roomId': roomId });
    } else {
        room_list[username] = { "roomId": roomId.toString(), 'topic': topic, 'tags': tags, 'rate': rate, 'free': true };
        callback({ 'response': 'Created room', 'roomId': roomId });
    }
}

exports.delete_room = function(username, callback) {
    if (room_list[username]) {
        delete room_list[username];
        callback({ 'response': 'Deleted room' });
    } else {
        callback({ 'response': 'Room is not exist!' });
    }
}

exports.enter_room = function(username, callback) {
    if (room_list[username]) {
        room_list[username]['free'] = false;
        callback({ 'response': 'A helper entered room' });
    } else {
        callback({ 'response': 'Room is not exist!' });
    }
}

exports.leave_room = function(username, callback) {
    if (room_list[username]) {
        room_list[username]['free'] = true;
        callback({ 'response': 'The helper left room' });
    } else {
        callback({ 'response': 'Room is not exist!' });
    }
}

exports.get_rooms_by_topic = function(id, topic, callback) {
    var result = [];
    room_list["John(mock user)"] = { 'topic': "Navigation", 'tags': "SFSU,student", 'rate': "3", 'roomId': '1234', 'free': true };
    room_list["Marry(mock user)"] = { 'topic': "Cooking", 'tags': "", 'rate': "4", 'roomId': '1235', 'free': true };
    // room_list["name3"] = {'topic': "Cooking", 'rate': "5", 'roomId': '1236', 'free': true};
    
    var count = 0;
    var isGeneral = (topic === "General");
    var friends = [];
    
    user.find({token: id}, function(err, users) {
        if (users.length != 0) {
            friends = users[0].friends;
            for (var i = 0; i < friends.length; i++) {
                var fname = friends[i];
                if (room_list[fname] && room_list[fname]['free'] && (isGeneral || room_list[fname]['topic'] === topic)) {
                    result.push({
                        'username': fname,
                        'topic': room_list[fname]['topic'],
                        'tags': room_list[fname]['tags'],
                        'rate': room_list[fname]['rate'],
                        'roomId': room_list[fname]['roomId']
                    });
                    count++;
                    console.log("find a friend");
                    if (count == 20) break;
                }
            }
        } else {
            console.log("Cannot find the user!");
        }
        
        for (var username in room_list) {
            if (count == 20) break;
            if (room_list[username]['free'] && (isGeneral || room_list[username]['topic'] === topic)) {
                if (friends.indexOf(username) != -1) continue;
                result.push({
                    'username': username,
                    'topic': room_list[username]['topic'],
                    'tags': room_list[username]['tags'],
                    'rate': room_list[username]['rate'],
                    'roomId': room_list[username]['roomId']
                });
                count++;
            }
        }
        
        callback({ 'response': 'Got List', 'size': count, 'list': JSON.stringify(result) });
    });
    
}
